### PR TITLE
#164139774 Create new recipe

### PR DESCRIPTION
**Pull request by blackdevelopa**
**What does this PR do?**
Create new recipe.

**Description of Task to be completed?**
- write unit test
- defined the recipe schema
- implement create article

**How should this be manually tested?**
- clone this repository
- run `npm install`
- start the dev server, 'npm run devstart`
- open up postman and generate an existing user's token using the endpoint, 'localhost:{PORT}/user/login. Ensure the email and password of the user is provided.
- add the token to the req.headers with 'authorization' as the key and the endpoint to 'localhost:{PORT}/recipe

**Any background context you want to provide?**
- New recipes can only be created by a registered user

**What are the relevant pivotal tracker stories?**
#164139774

**Screenshots (if appropriate)**

[#164139774]

### CODE REVIEW

File: server/controllers/recipe-controller.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+   * @returns {undefined}
+   */
+  static async create(req, res) {
+    const {
+      title,
+      ingredients,
+      steps,
+      cookingTime,
+      preparationTime,
+      tagList
```
Comment (andela-pessien) [inline server/controllers/recipe-controller.js]: This isn't your task :)

File: server/migrations/20190312121718-create-recipe.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        primaryKey: true,
+        type: Sequelize.INTEGER
+      },
+      title: {
+        type: Sequelize.STRING
+      },
+      slug: {
+        type: Sequelize.STRING
+      },
+      ingredients: {
```
Comment (andela-pessien) [inline server/migrations/20190312121718-create-recipe.js:16]: **[question]** Will we be storing the ingredients and their amounts together?

File: server/utils/initdb.js [hunk 3]
```diff
@@ -1,5 +1,8 @@
 import db from '../models';
+import envConfig from '../config/env-config';
 
-db.sequelize.sync().then(() => {
+const params = envConfig.NODE_ENV === 'test' ? { force: true } : {};
```
Comment (andela-pessien) [inline server/utils/initdb.js]: **[comment]** I don't expect us to keep syncing tables in production.

File: test/recipes.spec.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        .request(server)
+        .post(baseUrl)
+        .set({ accesstoken: token })
+        .send(recipe)
+        .end((err, res) => {
+          const { body } = res;
+          expect(res).to.have.status(201);
+          expect(body.recipe).to.be.an('object');
+          expect(body.recipe.id).to.be.a('number');
+          expect(body.recipe.title).to.be.a('string');
```
Comment (andela-pessien) [inline test/recipes.spec.js]: Shouldn't we test that the attributes returned match what was sent?

File: test/recipes.spec.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      delete recipe.title;
+      chai
+        .request(server)
+        .post(baseUrl)
+        .set({ accesstoken: token })
+        .send(recipe)
+        .end((err, res) => {
+          const { body } = res;
+          expect(res).to.have.status(404);
+          expect(body.errors).to.contain('Title must be present');
```
Comment (andela-pessien) [inline test/recipes.spec.js]: The error should match the offense :)

**Review by andela-pessien — CHANGES_REQUESTED** [review]

File: server/controllers/recipe-controller.js (hunk 1 above)
Comment (blackdevelopa) [reply server/controllers/recipe-controller.js]: Okay. Noted. I have fixed this.

File: test/recipes.spec.js (hunk 5 above)
Comment (blackdevelopa) [reply test/recipes.spec.js]: Yes. The code that solved for this was lost after my first rebase. This is fixed now.

File: server/migrations/20190312121718-create-recipe.js (hunk 2 above)
Comment (blackdevelopa) [reply server/migrations/20190312121718-create-recipe.js:16]: Yes. We will store the ingredients and the amounts together. An example will be, `1/2 cups of rice`. The ingredients will be an array of strings.

File: server/utils/initdb.js (hunk 3 above)
Comment (blackdevelopa) [reply server/utils/initdb.js]: We are working on this. This will be fixed in another PR.

File: test/recipes.spec.js (hunk 4 above)
Comment (blackdevelopa) [reply test/recipes.spec.js]: Okay. I will test that the response is same as the input.

File: server/migrations/20190312121718-create-recipe.js (hunk 2 above)
Comment (andela-pessien) [reply server/migrations/20190312121718-create-recipe.js:16]: 👍 We'll be further discussing this schema though as more features get built out

File: server/migrations/20190312121718-create-recipe.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      steps: {
+        type: Sequelize.JSONB
+      },
+      cookingTime: {
+        type: Sequelize.INTEGER
+      },
+      preparationTime: {
+        type: Sequelize.INTEGER
+      },
+      videoList: {
```
Comment (andela-pessien) [inline server/migrations/20190312121718-create-recipe.js:28]: We don't yet seem to be using this, plus wouldn't videos have the same challenges we highlighted with images?

File: server/models/recipe.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        validate: {
+          max: 50
+        }
+      },
+      slug: DataTypes.STRING,
+      ingredients: DataTypes.ARRAY(DataTypes.STRING),
+      steps: DataTypes.JSONB,
+      cookingTime: DataTypes.INTEGER,
+      preparationTime: DataTypes.INTEGER,
+      videoList: DataTypes.STRING
```
Comment (andela-pessien) [inline server/models/recipe.js]: Related to the above, would've thought this would be an array at least

File: server/models/recipe.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    },
+    {}
+  );
+  Recipe.associate = () => {
```
Comment (andela-pessien) [inline server/models/recipe.js]: How come the association between recipes and users isn't defined here?

File: server/utils/recipes.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+const recipeSchema = Joi.object().keys({
+  title: Joi.string()
+    .required()
+    .max(50),
+  ingredients: Joi.array()
+    .items(Joi.string().required(), Joi.string().required())
+    .required(),
+  steps: Joi.object()
+    .keys({
+      1: Joi.object().keys({
```
Comment (andela-pessien) [inline server/utils/recipes.js]: Only validates key `1`?

File: server/utils/slugify.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable no-useless-escape */
+const slugifyTitle = string => {
+  const a = 'àáäâãåăæçèéëêǵḧìíïîḿńǹñòóöôœṕŕßśșțùúüûǘẃẍÿź·/_,:;';
```
Comment (andela-pessien) [inline server/utils/slugify.js]: In this house we do not host single-letter variable names (except _maybe_ in for loops 🙂)

You should still be able to comfortably explain this code if we come back to it at the end of Simulations

**Review by andela-pessien — CHANGES_REQUESTED** [review]

File: server/migrations/20190312121718-create-recipe.js (hunk 6 above)
Comment (blackdevelopa) [reply server/migrations/20190312121718-create-recipe.js:28]: Not exactly, we will be using an array of strings instead. The intention is that the videos will belong to the recipe and not the steps.

File: server/models/recipe.js (hunk 7 above)
Comment (blackdevelopa) [reply server/models/recipe.js]: Yes. VideoList will be an array of strings. I have implemented this.

File: server/models/recipe.js (hunk 8 above)
Comment (blackdevelopa) [reply server/models/recipe.js]: Okay. This is valid. I will implement this right away

File: server/utils/slugify.js (hunk 10 above)
Comment (blackdevelopa) [reply server/utils/slugify.js]: I have updated this function.

**Review by andela-pessien — APPROVED** [review]
Looks good to me but please don't forget to reach a consensus concerning what happens to recipes of deleted users

**Pull request merged**